### PR TITLE
Common: added iglu-core Scala library (closes #135)

### LINKED ISSUE #135: Scala Core: add
Here's a proposal for scala-common library:

* Contains common data types using in all our applications and libraries working with Iglu, such as `SchemaVer`, `SchemaCriterion`, `SelfDescribingSchema` etc.
* Most common methods to work with these data types, such as `parse`, `validate`, `compare`, `generateIgluUri` etc.
* Some pieces (like regular expressions) can serve as reference implementation for Iglu on other platforms (JS, ObjC)
* Zero-dependency. It should be really small and slick and updated only when we're introducing some changes in related concepts (like "let's forbid underscores in Schema names"), not when new binary-incompatible scalaz is out. I think we can even omit a json4s dependency, all requiring implementation details, like parsing self-describing info can be injected via type classes or implicit classes.

We can roll it out in the near future, then use it in schema-guru, schema-ddl, iglu-scala-server, iglu static generator and only after that start to use it in iglu-scala-client. All these (and probably more) projects now have implemented various chunks of this functionality separately.

/cc @alexanderdean

### PR DESCRIPTION
**Pull request by chuwy**
This is ready for review.

This is mostly extraction from Iglu Scala Client, although I added some entities (like ordering) which can be useful at least in two Iglu-related projects, also some emphasis on comments, so it can be used as reference-implementation.

Also want to hear your thoughts on some design decisions (very minor I think):
1) Anything on [SelfDescribed](https://github.com/snowplow/iglu/compare/feature/iglu-core?expand=1#diff-aa8e8558e8b1224f4a30a4412e1e33a5R23) type class
2) [SchemaKey](https://github.com/snowplow/iglu/compare/feature/iglu-core?expand=1#diff-afa963f931416daec0b369219aad28c4R21). More precisely, what do you think about attaching it not only to Self-describing instances, but also to Self-describing Schemas. At first glance it contains all required data and satisfies requirements, but can be confusing, since by name it supposed to be attached only to instances. So it both handy and confusing.

/cc @fblundun

### CODE REVIEW

**Comment by chuwy** [comment]
3) Name. As we planned dir named scala-core, artifact - iglu-core. How I should refer to in docs? I'm using Iglu Scala Core.

File: 0-common/scala-core/vagrant/up.playbooks [hunk 1]
```diff
@@ -0,0 +1,3 @@
+oss-playbooks/java7.yml
+oss-playbooks/scala.yml
+oss-playbooks/sbt.yml
```
Comment (fblundun) [inline 0-common/scala-core/vagrant/up.playbooks]: Are we sure we want the iglu-core subproject to have its own Vagrant startup logic?

Comment (chuwy) [inline 0-common/scala-core/vagrant/up.playbooks]: I think this question is for @alexanderdean 
I'll be glad to remove it.

File: 0-common/scala-core/project/IgluCoreBuild.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  lazy val project = Project("iglu-core", file("."))
+    .settings(buildSettings: _*)
+    .settings(
+      libraryDependencies ++= Seq(
+        // Scala (test only)
+        Libraries.json4s,
+        Libraries.specs2
+      )
+    )
+}
```
Comment (fblundun) [inline 0-common/scala-core/project/IgluCoreBuild.scala]: I get this error when I try to `sbt compile`:

```
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	::          UNRESOLVED DEPENDENCIES         ::
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	:: org.scalaz.stream#scalaz-stream_2.10;0.7a: not found
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn] 	Note: Unresolved dependencies path:
[warn] 		org.scalaz.stream:scalaz-stream_2.10:0.7a
[warn] 		  +- org.specs2:specs2-common_2.10:3.3.1
[warn] 		  +- org.specs2:specs2-matcher_2.10:3.3.1
[warn] 		  +- org.specs2:specs2-core_2.10:3.3.1 (/vagrant/0-common/scala-core/project/IgluCoreBuild.scala#L30)
[warn] 		  +- com.snowplowanalytics:iglu-core:0.1.0-M1
```

This happens both in my host machine and in the Iglu VM I had made earlier. I haven't tried it inside a VM built using scala-core/Vagrantfile.

Comment (chuwy) [inline 0-common/scala-core/project/IgluCoreBuild.scala]: Weird. Why it even tries to get these for compile?

Comment (fblundun) [inline 0-common/scala-core/project/IgluCoreBuild.scala]: I'm not sure... The same thing happens when I make a new VM using scala-core/Vagrantfile. If I remove the specs2 dependency then the problem stops.

File: 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaKey.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+   * Usage:
+   * {{{
+   *   import com.snowplowanalytics.iglu.core.SchemaKey
+   *   implicit val schemaOrdering = SchemaKey.ordering
+   *   keys.sorted
+   * }}}
+   */
+  val ordering = Ordering.by { (key: SchemaKey) =>
+    (key.vendor, key.name, key.format, key.version.model, key.version.revision, key.version.addition)
+  }
```
Comment (fblundun) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaKey.scala]: I like the idea of a schema ordering. We could also eventually have:

* A similar ordering but for SchemaVers instead of SchemaKeys
* A scala.math.PartialOrdering for SchemaKeys where if two schemas have a different vendor, name, or format, neither of them is less than the other.

File: 0-common/scala-core/project/IgluCoreBuild.scala (hunk 2 above)
Comment (chuwy) [inline 0-common/scala-core/project/IgluCoreBuild.scala]: Ok, I've fixed it. But still wondering why sbt requires it for compile.

Comment (fblundun) [inline 0-common/scala-core/project/IgluCoreBuild.scala]: Cool, it's compiling for me now. Not sure where that unresolvable dependency came from.

**Comment by fblundun** [comment]
@chuwy your second point is interesting. Are you suggesting that a representation of for example the [link click schema](https://github.com/snowplow/iglu-central/blob/master/schemas/com.snowplowanalytics.snowplow/link_click/jsonschema/1-0-0) would include a schema key with the value `SchemaKey("com.snowplowanalytics.snowplow/link_click/jsonschema/1-0-0")`?

Or that it would it instead have the schema key for the schema for JSON schemas, i.e. `SchemaKey("com.snowplowanalytics.snowplow/schema/jsonschema/1-0-0")`?

Or am I misunderstanding?

**Comment by chuwy** [comment]
@fblundun I think I meant first one. But let me rephrase.

I'm wondering how useful/confusing it would be to use SchemaKey for both `schema` string key in self-describing instance and `self` object for self-describing Schemas.
For example, I was going to add following classes to iglu-core:

```scala
// Useful everywhere: scala-tracker, common-enrich, iglu-client
case class SelfDescribingInstance[JSON](key: SchemaKey, data: JSON)

// Useful particularly in schema guru, schema ddl, and probably in iglu-server
case class SelfDescribingSchema[JSON](key: SchemaKey, schema: JSON)
```
So, we could have both link_click Schema and link_click instance with same `SchemaKey`. And I found this situation both handy and confusing: we have same data (name, vendor, version) in two different (but related) contexts with two different representations (schema-string for instance and self-object for Schema).

Again, this may seem terribly minor, we could easily implement this simple classes in every application, but trying to make real reference-implementation I want to define core of Iglu (pun!) and its common entities.

**Comment by chuwy** [comment]
Updated classes.

**Comment by fblundun** [comment]
I think it's probably a good idea to reuse the SchemaKey class for schemas as well as instances. It means something different (for schemas the SchemaKey is a name, for instances the SchemaKey is a description of what the data is meant to look like). But it would be useful to do things like compare the key of a Schema with the key of an Instance before doing validation.

**Comment by fblundun** [comment]
Another thought. If this is a reference implementation, should we write an explicit standard for exactly what an implementation should include? Or does the standard just consist of every public field and method in the iglu-core source?

File: 0-common/scala-core/.travis.yml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+language: scala
+scala:
+   - 2.10.6
+   - 2.11.8
+jdk:
+  - oraclejdk7
+  - openjdk7
```
Comment (alexanderdean) [inline 0-common/scala-core/.travis.yml]: Can we replace openjdk7 with openjdk8?

File: 0-common/scala-core/CHANGELOG [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Version 0.1.0 (2016-xx-xx)
```
Comment (alexanderdean) [inline 0-common/scala-core/CHANGELOG]: There should only be a master CHANGELOG in the root - no child CHANGELOGs.

File: 0-common/scala-core/.travis.yml (hunk 4 above)
Comment (chuwy) [inline 0-common/scala-core/.travis.yml]: Not sure. @ninjabear mentioned there's no OpenJDK 8 in Travis CI. And seems this is true https://docs.travis-ci.com/user/languages/java, so I think we cannot.

File: 0-common/scala-core/project/BuildSettings.scala [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  lazy val basicSettings = Seq[Setting[_]](
+    organization       := "com.snowplowanalytics",
+    version            := "0.1.0-M1",
+    description        := "Core entities for Iglu",
+    scalaVersion       := "2.10.6",
+    crossScalaVersions := Seq("2.10.6", "2.11.8"),
+    scalacOptions      := Seq("-deprecation", "-encoding", "utf8", "-Yrangepos",
+                              "-feature", "-unchecked", "-Xlog-reflective-calls",
+                              "-Xlint"),
+    parallelExecution in Test := false, // Parallel tests cause havoc with LRU cache
```
Comment (alexanderdean) [inline 0-common/scala-core/project/BuildSettings.scala]: Parallel execution should be okay in test for this lib?

File: 0-common/scala-core/project/BuildSettings.scala [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    resolvers     ++= Dependencies.resolutionRepos
+  )
+
+  // Publish settings
+  // TODO: update with ivy credentials etc when we start using Nexus
+  lazy val publishSettings = Seq[Setting[_]](
```
Comment (alexanderdean) [inline 0-common/scala-core/project/BuildSettings.scala]: Replace with code for publishing to Bintray

File: 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaCriterion.scala [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+  /**
+   * Whether a SchemaKey is valid.
+   *
+   * It's valid if the vendor, name, format, and model all match
+   * and the supplied key's revision and addition do not exceed the
+   * criterion's revision and addition.
+   *
+   * This comparator will continue to function correctly when
+   * revisions are deprecated.
```
Comment (alexanderdean) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaCriterion.scala]: No longer plans to deprecate revisions, please remove comment

File: 0-common/scala-core/.travis.yml (hunk 4 above)
Comment (alexanderdean) [inline 0-common/scala-core/.travis.yml]: Okay - let's just remove openjdk7 then

File: 0-common/scala-core/project/Dependencies.scala [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    val json4s          = "3.3.0"
+    val specs2          = "3.3.1"
+  }
+
+  object Libraries {
+    // Scala (test only)
+    val json4s           = "org.json4s"                 %% "json4s-jackson"            % V.json4s          % "test"
+    val specs2           = "org.specs2"                 %% "specs2-core"               % V.specs2          % "test"
+  }
+}
```
Comment (alexanderdean) [inline 0-common/scala-core/project/Dependencies.scala]: Nicely minimal

File: 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaKey.scala (hunk 3 above)
Comment (alexanderdean) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaKey.scala]: This is nice. @fblundun can you create a ticket about the PartialOrdering idea?

File: 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaVer.scala [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+object SchemaVer {
+  /**
+   * Regular expression to validate SchemaVer
+   */
+  val schemaVerRegex = "\\d+-\\d+-\\d+".r
+
+  /**
+   * Regular expression to extract MODEL, REVISION, ADDITION
+   */
+  val modelRevisionAdditionRegex = "([0-9]+)-([0-9]+)-([0-9]+)".r
```
Comment (alexanderdean) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaVer.scala]: Didn't we have a user somewhere suggest a tighter definition for SchemaVer? As `007/1/020` is not a valid version...

File: 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaKey.scala (hunk 3 above)
Comment (chuwy) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaKey.scala]: I don't think its worth another ticket, I'll create it on next update.

File: 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaVer.scala (hunk 10 above)
Comment (chuwy) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SchemaVer.scala]: I think this one: https://github.com/snowplow/iglu-central/issues/288
I'll update it.

File: 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+ * Type-class representing essence of Self-describing data
+ * It can be instantiated for any type [[E]] which is potentially
+ * containing its own description (as [[SchemaKey]])
+ *
+ * @tparam E entity type, mostly intended for various JSON ADTs,
+ *           like Json4s, Jackson, circe, Argonaut etc,
+ *           but also can be anything that can bear reference to
+ *           its description like Thrift, Map[String, String] etc
+ */
+trait SelfDescribed[E] {
```
Comment (alexanderdean) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala]: Is there a reason why this is called `SelfDescribed` rather than `SelfDescribing`? `SelfDescribed` is past tense which doesn't quite fit for a trait...

File: 0-common/scala-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+  /**
+   * Example of [[SelfDescribed]] instance for usual case class
+   */
+  implicit object DescribedStringInstance extends SelfDescribed[DescribedString] {
+    def getSchemaKey(entity: DescribedString): Option[SchemaKey] =
+      Some(entity.key)
+  }
+
+}
```
Comment (alexanderdean) [inline 0-common/scala-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala]: This is really nice!

File: 0-common/scala-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+      (entity \ "schema") match {
+        case JString(schema) => SchemaKey.fromUri(schema)
+        case _               => None
+      }
+  }
+
+  /**
+   * Helper class bearing its Schema
+   */
+  case class DescribedString(key: SchemaKey, data: String)
```
Comment (alexanderdean) [inline 0-common/scala-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala]: What is a String about this case class? It looks like a `SelfDescribingJsonString` to me...

File: 0-common/scala-core/src/test/scala/com/snowplowanalytics/iglu/core/SchemaKeySpec.scala [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+// This library
+import IgluCoreCommon.DescribedString
+
+class SchemaKeySpec extends Specification { def is = s2"""
+  Specification for parsing SchemaKey
+    parse simple correct key from Iglu URI $e1
+    parse complex correct key from Iglu URI $e2
+    parse key with preceding zero $e3
+    fail to parse with missing REVISION $e4
+    fail to parse with invalid name $e5
```
Comment (alexanderdean) [inline 0-common/scala-core/src/test/scala/com/snowplowanalytics/iglu/core/SchemaKeySpec.scala]: Add a test for invalid SchemaVer (leading 0s)

File: 0-common/scala-core/vagrant/.gitignore [hunk 15]
```diff
@@ -0,0 +1,3 @@
+.peru
+oss-playbooks
+ansible
```
Comment (alexanderdean) [inline 0-common/scala-core/vagrant/.gitignore]: There should only be one master vagrant setup in the root, not child ones

File: 0-common/scala-core/vagrant/up.playbooks (hunk 1 above)
Comment (alexanderdean) [inline 0-common/scala-core/vagrant/up.playbooks]: Yes please, remove this child Vagrant. (Sounds rather Dickensian)

**Comment by alexanderdean** [comment]
`com.snowplowanalytics.iglu.core` is good. Scala Iglu Core sounds good.

**Comment by alexanderdean** [comment]
>  If this is a reference implementation, should we write an explicit standard for exactly what an implementation should include?

@fblundun can you create a ticket to explore this further in next release

**Comment by alexanderdean** [comment]
> I think it's probably a good idea to reuse the SchemaKey class for schemas as well as instances. It means something different (for schemas the SchemaKey is a name, for instances the SchemaKey is a description of what the data is meant to look like).

Agree. Luckily the SchemaKey terminology is relatively open-ended - if this were called SchemaUri it would be more of a stretch...

**Comment by alexanderdean** [comment]
That's a wrap from me! All minor things. This is looking great...

File: 0-common/scala-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala (hunk 13 above)
Comment (chuwy) [inline 0-common/scala-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala]: Actually, String here is meaningless placeholder. This case just shows an example that not only entities like JSON can bear description (`SchemaKey`) with it. More (or less) real-world example is `SchemaGuruOutput` case class which can be self-described instance if user specified `--vendor`, `--name` options and plain JSON Schema if he didn't.

File: 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala (hunk 11 above)
Comment (chuwy) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala]: Yes, there's a reason. My initial plan was to add two type classes: `SelfDescribing` and `SelfDescribed`. 

Below is kind of algebra of self-description. Long story short: `SelfDescribing` useful for creating Self-describing *something*, for example in trackers, and `SelfDescribed` useful for decomposing it, for example in Scala Common Enrich.

**`SelfDescribing`** here is for entities which we can describe. For example, we can add `SchemaKey` to JValue ```{"longitude": 32.3, "latitude": 65.1}```, so it will become Self-describing JSON instance (still JValue) and primary method for this type class would be:
```scala
def describe[E: SelfDescribed](schemaKey: SchemaKey, entity: E): E // context-bound isn't necessary
```
`E` here will always be JValue (or any other type which can be self-described), before `describe` call and after it. Also `describe`cannot fail. It can be double-described non-sense, but not failed (not an `Option[E]`) .

**`SelfDescribed`** on the other hand if for entities which already can bear `SchemaKey`. Primary method is:
```scala
 def getSchemaKey[E](entity: E): Option[SchemaKey]
```
which can fail. We also have no idea if `E` will remain same `E` (type, not value) without `SchemaKey` as in example with [SelfDescribedString](https://github.com/snowplow/iglu/pull/138/files/d91d7c2ddfc66e0f44196f73514f2aa2a5558686#diff-be297bb846e2436bb242db98395f3548R34). It also means single type class isn't sufficient.

I didn't add another type class because at that time I thought it could be over-engineering, but after short sight on Scala Tracker I changed my mind and I think it's worth to add both.

Comment (alexanderdean) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala]: Hey @chuwy - I *think* I get what you are saying. Notes:

* Please add the other type class so that the difference becomes clearer
* You talk about `SelfDescribing` in your second paragraph but then your example uses `E: SelfDescribed`. Is that a typo?
* When you add `describe`, it would be good to have an example in the test suite so it is easy to "get it"

Comment (chuwy) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala]: Nope, that was not a typo. 
If some type `E` have instance for type class `SelfDescribing`, we automatically can assume it can be `SelfDescribed`, but as I wrote in comment: this context-bound isn't necessary, I'm not planning to add it. its purpose was to make everything clearer... mission failed.

I'll add everything with tests and extensive comments.

Comment (alexanderdean) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala]: Right - got it. In that case, I think `describe` is not a great name for the function. I think you mean `makeSelfDecribing`. If I tell you to `describe` a `Dog("noisy")` I would expect to hear "the dog is noisy". If I ask you to `makeSelfDescribing` a `Dog("noisy")`, then I expect a `SelfDescribingDog("noisy")`.

Comment (chuwy) [inline 0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala]: Got it.

**Comment by Fokko** [comment]
Looks nice, what has to be done to get this merged?

**Comment by alexanderdean** [comment]
Hey @Fokko! This is actually scheduled into the next release. What are you using Iglu for?

**Comment by Fokko** [comment]
Nice! I would like to use the iglu to programmatically generate the Scala case classes, to make the decoding of the json files a bit easier and more consistent. So I could depend different application using `sbt` to import the schema's based on the json definitions.

**Comment by alexanderdean** [comment]
Hey @Fokko - that's a cool use case! Over time, we actually plan on letting Iglu auto-generate POJOs, case classes etc and host an internal Maven repository, to make this even easier: https://github.com/snowplow/iglu/issues/88

**Comment by Fokko** [comment]
Very nice. For thrift this is already quite easy, using the [SBT Scrooge](https://twitter.github.io/scrooge/SBTPlugin.html) plugin and works very well.

We don't use a full maven repository as SBT supports [GIT dependancies](http://www.scala-sbt.org/0.12.4/docs/Dormant/Full-Configuration.html), at the first run it will fetch the repository, compile the thrift files and add them to the class-path of the project.

Would be nice to have something like this for compiling iglu definition files to case-classes or POJO's.

**Comment by alexanderdean** [comment]
That's a good point re. git dependencies. Having Iglu host a Maven will have a couple of advantages:

1. Keeps Iglu the single source of truth (versus some apps using Iglu and others interfacing with the git repo backing Iglu directly)
2. Better support for handling SchemaVer within code artifacts. Our current thinking is that the MODEL will be part of the artifact name, and the REVISION and ADDITION (and perhaps PATCH) will be represented in the artifact version. This is because you may want to work with multiple MODELs in the same app, but would always want just the latest REVISION-ADDITION

**Comment by chuwy** [comment]
Superseded by #153.
